### PR TITLE
Feat: Adds interop with Arrow library using new method `Dataset.to_arrow()`

### LINKED ISSUE #204: 💡 Feature Request: Add `to_arrow()` on `Dataset` class
Arrow is a crucial library and support it natively is a big opportunity. So we can avoid to use to_pandas().

### PR DESCRIPTION
**Pull request by avirajsingh7**
Resolves #204 

As discuss in issue added a supoort pyarrow dataset instead of pyarrow table to handle a large datasets.

Supporting docs:
[pyarrow dataset ](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.dataset.html)
[pyarrow.Table.from_pandas](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.from_pandas)

Still need input on optimal chunksize...

## Summary by CodeRabbit

- **New Features**
  - Introduced Arrow dataset functionality for handling large tables by reading them in chunks.
  - Added `to_arrow` method for converting datasets to Arrow format with a specified chunk size.

- **Bug Fixes**
  - Ensured proper handling of large datasets by implementing chunk reading.

- **Tests**
  - Added new test case for verifying Arrow dataset functionality, including row and batch count checks.

### CODE REVIEW

File: airbyte/datasets/_base.py
Comment (avirajsingh7) [inline airbyte/datasets/_base.py]: What should be optimal chunk_size by default?

File: airbyte/caches/base.py
Comment (avirajsingh7) [inline airbyte/caches/base.py:152]: keeping return type as Any because this is return type of ds.dataset()

Comment (aaronsteers) [reply airbyte/caches/base.py:152]: I'd prefer to be explicit if we can. Can we use `pd.dataset` (or `Dataset`) as the return type?

File: airbyte/datasets/_base.py
Comment (aaronsteers) [reply airbyte/datasets/_base.py]: 10_000 or 100_000 feels like a decent place to start here. Maybe 100_000? And what do you think of adding a constant at the top of the file to hold this default?

Something like:

```py
DEFAULT_ARROW_CHUNKSIZE = 10)_000
```

And then the inline declaration here would be:

```suggestion
    def to_arrow_dataset(self, chunksize: int = DEFAULT_ARROW_CHUNKSIZE) -> Any:
```

Comment (aaronsteers) [inline airbyte/datasets/_base.py]: Regarding name here, I think we can abbreviate slightly, to simply `to_arrow()`. I think we can still use the more explicit `get_arrow_dataset()` at the Cache level, which is similar to the pairing of `Cache.get_pandas_dataframe()` and `Dataset.to_pandas()`.

File: airbyte/datasets/_sql.py
Comment (aaronsteers) [inline airbyte/datasets/_sql.py]: Ditto above. I think `to_arrow()` is sufficient here.

**Review by aaronsteers — CHANGES_REQUESTED** [review]
A few changes requested inline.

Also, can I request at least one test added? Feel free to replicate another test that validates `to_pandas()` capabilities. At minimum, we just want to make sure that the code can execute.

It is probably also worth adding a version of the test that forces a small chunk size, ensuring that the ">1 chunk" codepath is also working as expected.

File: airbyte/caches/base.py
Comment (avirajsingh7) [reply airbyte/caches/base.py:152]: Screenshot 2024-06-29 at 10 00 52 AM
Gone through official docs, return type can be Dataset

https://arrow.apache.org/docs/python/generated/pyarrow.dataset.dataset.html

File: airbyte/datasets/_base.py
Comment (avirajsingh7) [reply airbyte/datasets/_base.py]: 100_000 seems nice and added variable

**Comment by aaronsteers** [comment]
@coderabbitai review

**Comment by aaronsteers** [comment]
@avirajsingh7 - Auto-fix applied some lint fixing. And the other fixes you applied looked great. Lmk if this is ready for final review or if you are still applying additional changes.

**Comment by avirajsingh7** [comment]
@aaronsteers it is ready for review

File: tests/integration_tests/test_all_cache_types.py
Comment (aaronsteers) [inline tests/integration_tests/test_all_cache_types.py]: This looks great - thanks for adding! 🙏

**Review by aaronsteers — APPROVED** [review]
The changes look great. Thanks for the added changes and for the contribution. 🚀 

Will merge once all tests are passed.

File: airbyte/datasets/_base.py
Comment (aaronsteers) [reply airbyte/datasets/_base.py]: @avirajsingh7 - CodeRabbit found something I missed. Wdyt about adding a generic implementation here in the base class `DatasetBase.to_arrow()`?

File: airbyte/caches/base.py
Comment (aaronsteers) [reply airbyte/caches/base.py]: @avirajsingh7 - I think this one is okay. Resolving.

File: airbyte/datasets/_base.py
Comment (avirajsingh7) [reply airbyte/datasets/_base.py]: @aaronsteers I am not sure how this implementation should look like, can you help?, I am unable to think of any particular implementation we can have here.

Comment (avirajsingh7) [reply airbyte/datasets/_base.py]:
```
def to_arrow(self, chunksize: int) -> Dataset:
         return pyarrow.dataset.dataset([pyarrow.Table])
      
```

what do you think about this?

Comment (aaronsteers) [reply airbyte/datasets/_base.py]: @avirajsingh7 - Let's not worry about it for now, I think. We can always add additional implementations down the road.

**Comment by aaronsteers** [comment]
@avirajsingh7 - Looks like we have a warning here regarding `numpy` implementation:

<details><summary>Details</summary>
<p>

https://github.com/airbytehq/PyAirbyte/actions/runs/9726180729/job/26844327030?pr=281

image


</p>
</details> 

PyAirbyte treats warnings as fatal in our test suite, so we'd need to resolve the condition creating this warning or else add it to the list of ignored warnings in `pyproject.toml`.

I notices that NumPy 2.0 just came out and re-locking dependencies (or adding PyArrow) has bumped us to 2.0 version in `poetry.lock`. We could try adding a numpy constraint "<=2.0" if that is the source of the issue - but ignoring can also be fine. This [SO Answer](https://stackoverflow.com/a/40846742) seems to suggest the issue is fairly innocuous - while also giving some tips that might help resolve.

**Comment by avirajsingh7** [comment]
> @avirajsingh7 - Looks like we have a warning here regarding `numpy` implementation:
> 
> Details
> https://github.com/airbytehq/PyAirbyte/actions/runs/9726180729/job/26844327030?pr=281
> 
> image
> PyAirbyte treats warnings as fatal in our test suite, so we'd need to resolve the condition creating this warning or else add it to the list of ignored warnings in `pyproject.toml`.
> 
> I noticed that NumPy 2.0 just came out and re-locking dependencies (or adding PyArrow) has bumped us to 2.0 version in `poetry.lock`. We could try adding a numpy constraint "<=2.0" if that is the source of the issue - but ignoring can also be fine. This [SO Answer](https://stackoverflow.com/a/40846742) seems to suggest the issue is fairly innocuous - while also giving some tips that might help resolve.

@aaronsteers Whatever you recommend, I can go with that,
we can try adding constraint, if adding constraint solve this well and good, Else it is innocuous we can also ignore this ([Gpt Response](https://chatgpt.com/share/1502969d-d868-4482-8529-6a94ce715976))

**Comment by aaronsteers** [comment]
@avirajsingh7 - On the tests side, I'm seeing this weird error. I'm not sure if it is new, since pulling in version updates from `main`, or if this was here before and we missed it.

```
UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
```

https://github.com/airbytehq/PyAirbyte/actions/runs/9832269420

**Comment by aaronsteers** [comment]
BigQuery is failing at the `pandas_chunks = pd.read_sql_table(`.

To merge, we may want to declare BigQuery unsupported and simply raise `NotImplementedError` with a clear error message for BigQuery. @avirajsingh7 - Lmk if you are able to pick this up.

**Comment by avirajsingh7** [comment]
> BigQuery is failing at the `pandas_chunks = pd.read_sql_table(`.
> 
> To merge, we may want to declare BigQuery unsupported and simply raise `NotImplementedError` with a clear error message for BigQuery. @avirajsingh7 - Lmk if you are able to pick this up. We are very close!
> 
> > `FAILED tests/integration_tests/test_all_cache_types.py::test_faker_read[BigQuery] - sqlalchemy.exc.InvalidRequestError: Could not reflect: requested table(s) not available in Engine(bigquery://dataline-integration-testing?credentials_path=%2Ftmp%2Ftmpt3yt8wxn.txt) schema 'test_deleteme_fx14pe': (users)`

@aaronsteers I think this fix should work,
is there any way I can get which cache we are using so we can skip unit test for big query but keep using them for other cache

**Comment by aaronsteers** [comment]
@avirajsingh7 - Thanks for taking the revision for the BigQuery issue. I did some digging and found this issue from a while back, describing a similar problem for `to_pandas()` generally.

- https://github.com/airbytehq/PyAirbyte/issues/165

File: tests/integration_tests/test_all_cache_types.py
Comment (aaronsteers) [inline tests/integration_tests/test_all_cache_types.py]:
```suggestion
    if "bigquery" not in new_generic_cache.get_sql_alchemy_url():
        # BigQuery doesn't support to_arrow
        # https://github.com/airbytehq/PyAirbyte/issues/165
        arrow_dataset = result["users"].to_arrow(max_chunk_size=10)
```

File: tests/integration_tests/test_all_cache_types.py
Comment (aaronsteers) [inline tests/integration_tests/test_all_cache_types.py]:
```suggestion
        assert sum(1 for _ in arrow_dataset.to_batches()) == FAKER_SCALE_A / 10
```

File: airbyte/caches/bigquery.py
Comment (aaronsteers) [inline airbyte/caches/bigquery.py]:
```suggestion
        """Raises NotImplementedError; BigQuery doesn't support `pd.read_sql_table`.
 
        https://github.com/airbytehq/PyAirbyte/issues/165
        """
```

File: airbyte/caches/bigquery.py
Comment (aaronsteers) [reply airbyte/caches/bigquery.py]:
```suggestion
        """Raises NotImplementedError; BigQuery doesn't support `pd.read_sql_table`.
 
        For more details, see:
        - https://github.com/airbytehq/PyAirbyte/issues/165
        """
```

**Comment by aaronsteers** [comment]
🚀 🙌

**Comment by avirajsingh7** [comment]
Thanks @aaronsteers for all your efforts and help

**Pull request merged**
